src/sst/core/serialization/impl/serialize_array.h

```diff
@@ -20,200 +20,197 @@
-{
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::ser_array_wrapper<T, IntType> arr, serializer& ser)
-    {
-        ser.primitive(arr.sizeptr);
-        for ( int i = 0; i < arr.sizeptr; i++ ) {
-            ser& arr[i];
-        }
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<ELEM_T (*)[SIZE]> : pvt::serialize_impl_fixed_array<ELEM_T (*)[SIZE], ELEM_T, SIZE>
+{};
 
-        // TODO: Implement mapping mode
-    }
-};
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<std::array<ELEM_T, SIZE>*> :
+    pvt::serialize_impl_fixed_array<std::array<ELEM_T, SIZE>*, ELEM_T, SIZE>
+{};
 
-/**
-   Version of serialize that works for statically allocated arrays of
-   void*.
- */
-template <class IntType>
-class serialize_impl<pvt::ser_array_wrapper<void, IntType>>
+// Serialize dynamic arrays
+template <typename ELEM_T, typename SIZE_T>
+struct serialize_impl<pvt::array_wrapper<ELEM_T, SIZE_T>>
 {
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::ser_array_wrapper<void, IntType> arr, serializer& ser)
+    void operator()(pvt::array_wrapper<ELEM_T, SIZE_T>& ary, serializer& ser)
     {
-        ser.binary(arr.bufptr, arr.sizeptr);
-
-        // TODO: Implement mapping mode
+        switch ( const auto mode = ser.mode() ) {
+        case serializer::MAP:
+            // TODO: Implement mapping mode
+            // Functions like pvt::serialize_array_map() and pvt::new_fixed_array_map() should be used to reduce code
+            // size
+            break;
+
+        default:
+            if constexpr ( std::is_void_v<ELEM_T> || pvt::is_trivial_element_v<ELEM_T> )
+                ser.binary(ary.ptr, ary.size);
+            else {
+                ser.primitive(ary.size);
+                if ( mode == serializer::UNPACK ) ary.ptr = new ELEM_T[ary.size];
+                pvt::serialize_array(ser, ary.ptr, ary.size, pvt::serialize_array_element<ELEM_T>);
+            }
+            break;
+        }
     }
 };
 
-/***   Other Specializations (raw_ptr and trivially_serializable)  ***/
-
 /**
    Version of serialize that works for copying raw pointers (only
    copying the value of the pointer.  Note that this is only useful
    if the value is going to be sent back to the originator, since it
    won't be valid on the other rank.
  */
-template <class TPtr>
-class serialize_impl<pvt::raw_ptr_wrapper<TPtr>>
+template <typename ELEM_T>
+struct serialize_impl<pvt::raw_ptr_wrapper<ELEM_T>>
 {
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::raw_ptr_wrapper<TPtr> ptr, serializer& ser)
+    void operator()(pvt::raw_ptr_wrapper<ELEM_T>& a, serializer& ser)
     {
-        ser.primitive(ptr.bufptr);
-
-        // TODO: Implement mapping mode
+        switch ( ser.mode() ) {
+        case serializer::MAP:
+            // TODO: Implement mapping mode
```

Comment:
Mapping mode does not not need to be implemented for raw pointers. If it's being serialized as just the value of the pointer, then we don't want to follow it to map the data it points to.

Reply:
I can remove that comment then and clean up the code. The previous code had a `TODO` about mapping mode in the raw pointers serialization, but I wasn't sure what needed to be done.